From db422a36f4b964e9336c0229ebb2b79ada3fc207 Mon Sep 17 00:00:00 2001
From: yury <yury.korolev@gmail.com>
Date: Tue, 24 Aug 2010 23:31:48 +0400
Subject: [PATCH] Micro optimization for build_named_route_call in PolymorphicRoutes:

1. use map instead of inject
2. use [].join("_") instead of '<<'. It is a little bit faster for ruby 1.9.2 and x2 faster for ruby 1.8.7. http://gist.github.com/548143
---
 .../action_dispatch/routing/polymorphic_routes.rb  |   20 ++++++++++----------
 1 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/actionpack/lib/action_dispatch/routing/polymorphic_routes.rb b/actionpack/lib/action_dispatch/routing/polymorphic_routes.rb
index 31dba83..142cd08 100644
--- a/actionpack/lib/action_dispatch/routing/polymorphic_routes.rb
+++ b/actionpack/lib/action_dispatch/routing/polymorphic_routes.rb
@@ -148,29 +148,29 @@ module ActionDispatch
         def build_named_route_call(records, inflection, options = {})
           unless records.is_a?(Array)
             record = extract_record(records)
-            route  = ''
+            route  = []
           else
             record = records.pop
-            route = records.inject("") do |string, parent|
+            route = records.map do |parent|
               if parent.is_a?(Symbol) || parent.is_a?(String)
-                string << "#{parent}_"
+                parent
               else
-                string << ActiveModel::Naming.plural(parent).singularize
-                string << "_"
+                ActiveModel::Naming.plural(parent).singularize
               end
             end
           end
 
           if record.is_a?(Symbol) || record.is_a?(String)
-            route << "#{record}_"
+            route << record
           else
             route << ActiveModel::Naming.plural(record)
-            route = route.singularize if inflection == :singular
-            route << "_"
-            route << "index_" if ActiveModel::Naming.uncountable?(record) && inflection == :plural
+            route = [route.join("_").singularize] if inflection == :singular
+            route << "index" if ActiveModel::Naming.uncountable?(record) && inflection == :plural
           end
 
-          action_prefix(options) + route + routing_type(options).to_s
+          route << routing_type(options)
+
+          action_prefix(options) + route.join("_")
         end
 
         def extract_record(record_or_hash_or_array)
-- 
1.7.1

